### PR TITLE
kernel: shim-eth-vlan: fixed bug in update and restore qdisc, adding also support for devices with multiple tx queues

### LINKED ISSUE #734: BUG: kernel NULL pointer, update_qdisc+0x23/0x130
Branch: pristine-1.3

This error occurs once you ctrl-c the ipcm and then start it again with the same configuration. Seems to be effected in configurations which uses the shim_eth_vlan module.

```
rina-shim-eth-vlan(DBG): [create_vlan_interface_name] Complete vlan-interface-name = 'eth19.100'
Sep 14 09:54:27 alpha kernel: [  392.236971] BUG: unable to handle kernel NULL pointer dereference at 0000000000000018
Sep 14 09:54:27 alpha kernel: [  392.237107] IP: [<ffffffffa070c703>] update_qdisc+0x23/0x130 [shim_eth_vlan]
Sep 14 09:54:27 alpha kernel: [  392.237223] PGD dab10067 PUD df9ab067 PMD 0
Sep 14 09:54:27 alpha kernel: [  392.237306] Oops: 0000 [#1] SMP
Sep 14 09:54:27 alpha kernel: [  392.237370] Modules linked in: shim_eth_vlan normal_ipcp rina_default_plugin rinarp arp826 rina_personality_default 8021q garp stp mrp llc nfsv3 ib_iser rdma_cm iw_cm ib_cm ib_sa ib_mad ib_core ib_addr iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc loop ohci_pci ohci_hcd snd_pcm ehci_pci ehci_hcd radeon kvm_amd kvm ttm drm_kms_helper drm psmouse nv_tco usbcore snd_timer snd soundcore shpchp k8temp serio_raw i2c_nforce2 i2c_algo_bit amd64_edac_mod i2c_core edac_mce_amd edac_core pcspkr evdev parport_pc parport usb_common processor thermal_sys button ext4 crc16 mbcache jbd2 dm_mod sg sd_mod crc_t10dif crct10dif_common ata_generic floppy pata_amd sata_nv libata e1000e ptp pps_core forcedeth scsi_mod [last unloaded: shim_eth_vlan]
Sep 14 09:54:27 alpha kernel: [  392.238995] CPU: 3 PID: 4838 Comm: ipcm Not tainted 3.15.0-pristine-1.3 #7
Sep 14 09:54:27 alpha kernel: [  392.239097] Hardware name: Supermicro H8DMR-82/H8DMR-82, BIOS 080014  04/01/2008
Sep 14 09:54:27 alpha kernel: [  392.239207] task: ffff8800da73b5b0 ti: ffff8800db5f8000 task.ti: ffff8800db5f8000
Sep 14 09:54:27 alpha kernel: [  392.239317] RIP: 0010:[<ffffffffa070c703>]  [<ffffffffa070c703>] update_qdisc+0x23/0x130 [shim_eth_vlan]
Sep 14 09:54:27 alpha kernel: [  392.239464] RSP: 0018:ffff8800db5fba10  EFLAGS: 00010282
Sep 14 09:54:27 alpha kernel: [  392.239543] RAX: 0000000000000000 RBX: ffff88007b350000 RCX: 0000000000000000
Sep 14 09:54:27 alpha kernel: [  392.239648] RDX: ffff88011bd2fb60 RSI: ffffffffa0711190 RDI: ffff88007b350000
Sep 14 09:54:27 alpha kernel: [  392.239752] RBP: 000000000000000f R08: 0000000000000000 R09: 000000000000fffe
Sep 14 09:54:27 alpha kernel: [  392.239856] R10: 0000000000000000 R11: 0000000000000582 R12: ffff88011bd2fb60
Sep 14 09:54:27 alpha kernel: [  392.239961] R13: ffff88011bd2f800 R14: ffffffffa071005e R15: ffff88011c795a00
Sep 14 09:54:27 alpha kernel: [  392.240014] FS:  00007febfa7ec740(0000) GS:ffff88011fc80000(0000) knlGS:0000000000000000
Sep 14 09:54:27 alpha kernel: [  392.240014] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Sep 14 09:54:27 alpha kernel: [  392.240014] CR2: 0000000000000018 CR3: 00000000df876000 CR4: 00000000000007e0
Sep 14 09:54:27 alpha kernel: [  392.240014] Stack:
Sep 14 09:54:27 alpha kernel: [  392.240014]  ffffffffa070b0b4 ffff8800dfaa6fc0 000000000000000f ffff88011bd2fb60
Sep 14 09:54:27 alpha kernel: [  392.240014]  ffff88011bd2f800 ffffffffa070ce44 ffff8800df152e30 00000064a06d02f0
Sep 14 09:54:27 alpha kernel: [  392.240014]  ffff8800db5fbb00 ffff8800db440b00 0000000000000001 ffff88011c5d9c40
Sep 14 09:54:27 alpha kernel: [  392.240014] Call Trace:
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffffa070b0b4>] ? create_vlan_interface_name+0xe4/0xf0 [shim_eth_vlan]
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffffa070ce44>] ? eth_vlan_assign_to_dif+0x234/0x516 [shim_eth_vlan]
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffffa06cb1fb>] ? notify_ipcp_assign_dif_request+0xcb/0x190 [rina_personality_default]
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffffa06cb130>] ? notify_ipcp_allocate_flow_response+0x210/0x210 [rina_personality_default]
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff814fd373>] ? dispatcher+0x153/0x1f0
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff814437c3>] ? genl_family_rcv_msg+0x193/0x360
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81443990>] ? genl_family_rcv_msg+0x360/0x360
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81443a09>] ? genl_rcv_msg+0x79/0xc0
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81443100>] ? netlink_rcv_skb+0xa0/0xc0
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81443614>] ? genl_rcv+0x24/0x40
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8144274c>] ? netlink_unicast+0xfc/0x190
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81442b0e>] ? netlink_sendmsg+0x32e/0x770
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff813fbc4d>] ? sock_sendmsg+0x8d/0xc0
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8109bff1>] ? check_preempt_wakeup+0x151/0x1d0
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8119635e>] ? __mem_cgroup_commit_charge+0x7e/0x390
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff813fb989>] ? move_addr_to_kernel.part.19+0x19/0x60
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff813fc0d8>] ? ___sys_sendmsg+0x408/0x420
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8136740c>] ? pty_write+0x4c/0x50
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff810a46d2>] ? remove_wait_queue+0x12/0x50
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8151b780>] ? __do_page_fault+0x220/0x520
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81098816>] ? set_next_entity+0x56/0x70
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8109fab8>] ? pick_next_task_fair+0x6c8/0x820
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff81011616>] ? __switch_to+0x196/0x4f0
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff813fce7e>] ? __sys_sendmsg+0x3e/0x80
Sep 14 09:54:27 alpha kernel: [  392.240014]  [<ffffffff8152076d>] ? system_call_fastpath+0x1a/0x1f
Sep 14 09:54:27 alpha kernel: [  392.240014] Code: fd ff ff 0f 1f 44 00 00 66 66 66 66 90 41 55 41 54 48 c7 c6 90 11 71 a0 55 53 48 89 fb 49 89 d4 48 83 ec 08 48 8b 87 90 03 00 00 <48> 8b 78 18 48 83 c7 10 e8 50 b2 de e0 85 c0 89 c5 75 12 48 83
Sep 14 09:54:27 alpha kernel: [  392.240014] RIP  [<ffffffffa070c703>] update_qdisc+0x23/0x130 [shim_eth_vlan]
Sep 14 09:54:27 alpha kernel: [  392.240014]  RSP <ffff8800db5fba10>
Sep 14 09:54:27 alpha kernel: [  392.240014] CR2: 0000000000000018
Sep 14 09:54:27 alpha kernel: [  392.313835] ---[ end trace c6ac4a3b7533eda0 ]---
```

Trying to remove the module after this error does not work and leads to ipcm start failure with the following message:
```
5075(1442217924)#librina.concurrency (CRIT): Cannot destroy mutex
```

Removing all the rina modules from the kernel and trying to load them again does not work and generate the same error as seen here before. The only way i found is to restart the node.

### PR DESCRIPTION
**Pull request by edugrasa**
Fixes #734; fixed bug in update and restore qdisc, adding also support for devices with multiple tx queues

### CODE REVIEW

**Comment by edugrasa** [comment]
Maintainers: @sandervrijders 
It would be great if you could ACK it quickly since the bug it fixes is a bit annyoing (after destroying a shim-eth-vlan IPCP the system will have to be rebooted in order to be usable again)

File: linux/net/rina/ipcps/shim-eth-vlan.c [hunk 1]
```diff
@@ -956,31 +956,41 @@ int update_qdisc(struct net_device *    dev,
...
-	dev->qdisc = sch;
+	/* For each TX queue, create a new shim-eth-qdisc instance, attach
+	 * it to the queue and then destroy the old qdisc
+	 */
+	for (q_index = 0; q_index < dev->num_tx_queues; q_index++) {
+		queue = netdev_get_tx_queue(dev, q_index);
+		qdisc = qdisc_create_dflt(queue, &shim_eth_qdisc_ops, 0);
+		if (!qdisc) {
+			LOG_ERR("Problems creating shim-eth-qdisc");
+			continue;
```
Comment (sandervrijders) [inline linux/net/rina/ipcps/shim-eth-vlan.c]: Shouldn't it bail out here?

Comment (edugrasa) [inline linux/net/rina/ipcps/shim-eth-vlan.c]: Right, updated code with a new commit

Comment (sandervrijders) [inline linux/net/rina/ipcps/shim-eth-vlan.c]: Ok. Have a look at git commit --amend. Will reduce the number of commits.
On Sep 15, 2015 9:55 AM, Eduard Grasa <notifications@github.com> wrote:In linux/net/rina/ipcps/shim-eth-vlan.c:
>  	if (dev->flags & IFF_UP)
>  		dev_deactivate(dev);
>  
> -	dev_graft_qdisc(netdev_get_tx_queue(dev, 0), sch);
> -	dev->qdisc = sch;
> +	/* For each TX queue, create a new shim-eth-qdisc instance, attach
> +	 * it to the queue and then destroy the old qdisc
> +	 */
> +	for (q_index = 0; q_index < dev->num_tx_queues; q_index++) {
> +		queue = netdev_get_tx_queue(dev, q_index);
> +		qdisc = qdisc_create_dflt(queue, &shim_eth_qdisc_ops, 0);
> +		if (!qdisc) {
> +			LOG_ERR("Problems creating shim-eth-qdisc");
> +			continue;

Right, updated code with a new commit

**Comment by vmaffione** [comment]
ACK, usual tests work.

File: linux/net/rina/ipcps/shim-eth-vlan.c (hunk 1 above)
Comment (edugrasa) [inline linux/net/rina/ipcps/shim-eth-vlan.c]: Ok, thanks, will use it in forthcoming pull requests.

**Comment by sandervrijders** [comment]
ACK

**Comment by sandervrijders** [comment]
Please fix your clock, these commits are in the future.

**Comment by vmaffione** [comment]
Usual tests work --> ACK.

**Comment by kewinrausch** [comment]
Well, the ipcm is loaded again without problem, but i got these messages in the kernel log:
```
Sep 16 03:08:27 alpha kernel: [  800.738122] rina-shim-eth-vlan(ERR): [update_qdisc] qdisc not found on device eth1
Sep 16 03:08:27 alpha kernel: [  800.738507] rina-shim-eth-vlan(ERR): [eth_vlan_assign_to_dif] Problems creating queue discipline
Sep 16 03:08:27 alpha kernel: [  800.738915] rina-kipcm(ERR): [notify_ipcp_assign_dif_request] Assign to dif 100/// operation failed for IPC process 1
```
are they serious ones? I guess yes

**Comment by vmaffione** [comment]
Yes, it's serious. Didn't see anything like that in my tests

**Comment by kewinrausch** [comment]
This is the output of list-ipcps:
```
IPCM >>> Current IPC processes (id | name | type | state | Registered applications | Port-ids of flows provided)
    1 | shim-ipc:1:: | shim-eth-vlan | INITIALIZED | - | -
    2 | alpha.lim:1:: | normal-ipc | ASSIGNED TO DIF lim | alpha.dc-1-- | -
    3 | alpha.dc:1:: | normal-ipc | ASSIGNED TO DIF dc | - | -
```
Did you succeed in perform correctly enrollment + communication after this patch application?
I'm checking if my repo is aligned well, to be sure I'm not saying something bad.

**Comment by edugrasa** [comment]
Yes on the VMs, I guess this happened in the Vwall?

**Comment by kewinrausch** [comment]
The problem happened in the VM, but now it's resolved. 
It was just a problem with git alignment.

Late ACK also from me. :(

**Pull request merged**